### PR TITLE
treesitter: allow to iterate over node children

### PR DESCRIPTION
**Pull request by vigoux**
Really simple PR just to provide a way to simply iterate over a nodes children in lua (plus being able to extract field names with this).

The iterator allows things like this :

```lua
-- Recursively prints a subtree
function print_tree(root, indent)
  local indentation = indent or ""
  for node, field in root:iter_children() do
    if node:named() then
      if field then
        print(indentation, field, ":", node:type())
      else
        print(indentation, node:type())
      end
      print_tree(node, indentation .. "  ")
    end
  end
end
```

Also adds a new node method to get a child node by field.

### CODE REVIEW

**Comment by vigoux** [comment]
The openbsd error seems not related to the PR, can the build be restarted ?

**Review by teto — APPROVED** [review]

File: src/nvim/lua/treesitter.c [hunk 1]
```diff
@@ -646,6 +654,36 @@ static int node_symbol(lua_State *L)
...
+  const char *field_name = luaL_checklstring(L, 2, &name_len);
+
+  TSTreeCursor cursor = ts_tree_cursor_new(node);
+
+  lua_newtable(L);  // [table]
+  unsigned int curr_index = 0;
+
+  if (ts_tree_cursor_goto_first_child(&cursor)) {
+    // TODO(vigoux): It seems that multiple fields with the same name
+    //  are possible, but is that really a good practice ?
```
Comment (tjdevries) [inline src/nvim/lua/treesitter.c:674]: Do we need this TODO here?

If you're sure you only need one, we could optionally pass in another variable (`find_first`) and then it doesn't return a list, it just returns the node. And it quits out earlier.

Comment (vigoux) [reply src/nvim/lua/treesitter.c:674]: Maybe we don't need it but I find it weird to hav multiple node with the same field

Comment (theHamsta) [reply src/nvim/lua/treesitter.c]: You can just return multiple values, then only callers looking for more values would see them or capturing them in a table would see them.

File: runtime/doc/lua.txt [hunk 2]
```diff
@@ -622,6 +622,15 @@ Node methods						*lua-treesitter-node*
 tsnode:parent()						*tsnode:parent()*
 	Get the node's immediate parent.
 
+tsnode:iter_children()					*tsnode:iter_children()*
+	Iterates over all the direct children of {tsnode}, regardless of
+	wether they are named or not.
```
Comment (brcolow) [inline runtime/doc/lua.txt:627]: I know this was merged but small typo: "wether" => "whether"

Comment (vigoux) [reply runtime/doc/lua.txt:627]: Oops, this will be fixed in a later pr, thanks for reporting!

**Pull request merged**